### PR TITLE
A fully functional reseting password

### PR DESCRIPTION
**Pull request by stevenxzhou**
Hi Team members, 

I've followed the comments you guys did last time and I've made all the necessary changes.

However, there might still be some minor issues there I am not aware of. Please do another review and instruct me to make it more perfect. 

Thank you,
Steven

### CODE REVIEW

File: config.js [hunk 1]
```diff
@@ -396,8 +396,8 @@ config.email = {
     'smtpTransport': {
         'service': 'Gmail',
         'auth': {
-            'user': 'my.email@gmail.com',
-            'pass': 'myemailpassword'
+            'user': 'taybson@gmail.com',
```
Comment (stuartf) [inline config.js]: Looks like you checked in your `config.js`, you should change your gmail password as it's now been shown to the world for a month.

Comment (taybson) [inline config.js]: @stuartf Haha, I'll  do the change, thanks for reminding me about that! : )

**Comment by simong** [comment]
For what it's worth, I have quite a bit of follow-up sitting in my [resetpw branch](https://github.com/simong/Hilary/tree/resetp)w. It doesn't look like I'll have time in the short run to go through this, but it might be useful to someone else?

File: node_modules/oae-authentication/emailTemplates/reset.html.jst [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
+        <meta name="viewport" content="width=device-width" />
+        <style>
+
+            /*************
+             * CORE CSS **
+             *************/
+
+            body {
+                background-color: <%= skinVariables['body-background-color'] %>;
```
Comment (sallakarppinen) [inline node_modules/oae-authentication/emailTemplates/reset.html.jst]: Kudos for using the skinning variables in the email. :+1:

File: node_modules/oae-authentication/lib/api.js [hunk 3]
```diff
@@ -16,23 +16,25 @@
 var _ = require('underscore');
 var events = require('events');
 var passport = require('passport');
-
+var crypto = require('crypto');
```
Comment (sallakarppinen) [inline node_modules/oae-authentication/lib/api.js]: This needs to be added to package.json since I don't think we use it already.

File: node_modules/oae-authentication/lib/api.js [hunk 4]
```diff
@@ -677,6 +679,113 @@ var getUserIdFromLoginId = module.exports.getUserIdFromLoginId = function(tenant
...
+            // No user found with that login id
+            return callback({'code': 401, 'msg': 'No password found for this principal'});
+        }
+
+        // Get the secret column out of the row
+        var dbSecret = rows[0].get('secret');
+
+        // If the secret column was not found or its value didn't match, something is wrong.
+        if (!dbSecret || dbSecret.value !== secret) {
+            log().warn({'loginId': loginId}, 'Potential hacking attempt detected');
```
Comment (sallakarppinen) [inline node_modules/oae-authentication/lib/api.js]: This logging message would be more useful if it had more information about what actually happened - there are many reasons why a secret could not be found that are not related to hacking (for example a user clicked on an old link).

File: node_modules/oae-authentication/lib/api.js [hunk 5]
```diff
@@ -677,6 +679,113 @@ var getUserIdFromLoginId = module.exports.getUserIdFromLoginId = function(tenant
...
+            return callback({'code': 401, 'msg': 'The secret was not found or incorrect'});
+			
+		    // If the secret column was found.
+        }else{
```
Comment (sallakarppinen) [inline node_modules/oae-authentication/lib/api.js]: Formatting - should have spaces.

File: node_modules/oae-authentication/lib/rest.js [hunk 6]
```diff
@@ -44,3 +43,93 @@ require('./strategies/twitter/rest');
...
+ * @HttpResponse                400                     You cannot mix clearing an entire element and an optionalKey
+ * @HttpResponse                401                     Only authorized tenant admins can change config values
+ * @HttpResponse                401                     User is not allowed to update config value ...
+ * @HttpResponse                404                     Config value ... does not exist
+ */
+OAE.tenantRouter.on('get', '/api/auth/local/reset/init/:username', _getResetPasswordSecret);
+
+
+var _resetPassword = function(req, res){
+	process.stdout.write("hello: ");
```
Comment (sallakarppinen) [inline node_modules/oae-authentication/lib/rest.js]: I think this was left in by mistake... :)

File: node_modules/oae-authentication/lib/rest.js [hunk 7]
```diff
@@ -44,3 +43,93 @@ require('./strategies/twitter/rest');
...
+
+var _getResetPasswordSecret = function(req, res){
+
+	AuthenticationAPI.getResetPasswordSecret(req.ctx, req.params.tenantAlias, req.params.username, function(err, token){
+	    if (err) {
+	        return callback(err);
+	    }
+
+		return res.send(token);
+	})
```
Comment (sallakarppinen) [inline node_modules/oae-authentication/lib/rest.js]: Missing semicolon.

File: node_modules/oae-authentication/lib/rest.js [hunk 8]
```diff
@@ -44,3 +43,93 @@ require('./strategies/twitter/rest');
...
+ * @Return      {string}
+ * @HttpResponse                200                     Configuration value cleared
+ * @HttpResponse                400                     Missing configuration. Example configuration: `"oae-authentication/twitter/enabled"`
+ * @HttpResponse                400                     Missing tenant alias
+ * @HttpResponse                400                     You cannot mix clearing an entire element and an optionalKey
+ * @HttpResponse                401                     Only authorized tenant admins can change config values
+ * @HttpResponse                401                     User is not allowed to update config value ...
+ * @HttpResponse                404                     Config value ... does not exist
+ */
+OAE.globalAdminRouter.on('post', '/api/auth/local/reset/init/:tenantAlias/:userId', _getResetPasswordSecret);
```
Comment (sallakarppinen) [inline node_modules/oae-authentication/lib/rest.js]: Are we sure we want to add this functionality for admin users as well as regular users?

File: node_modules/oae-authentication/lib/rest.js [hunk 9]
```diff
@@ -44,3 +43,93 @@ require('./strategies/twitter/rest');
...
+
+var _resetPassword = function(req, res){
+	process.stdout.write("hello: ");
+	AuthenticationAPI.resetPassword(req.ctx, req.params.tenantAlias, req.params.username, req.body.secret, req.body.newPassword, function(err, changed) {
+	    if (err) {
+	        return callback(err);
+	    }
+		
+		return res.send(200);
+	})
```
Comment (sallakarppinen) [inline node_modules/oae-authentication/lib/rest.js]: Missing semicolon here as well.

File: node_modules/oae-authentication/lib/api.js (hunk 3 above)
Comment (brecke) [inline node_modules/oae-authentication/lib/api.js]: `npm install --save crypto` adds the package to `package.json` directly beside installing it.

File: node_modules/oae-authentication/lib/rest.js (hunk 8 above)
Comment (brecke) [inline node_modules/oae-authentication/lib/rest.js]: Probably not. We should have that removed.

**Comment by sallakarppinen** [comment]
`config.js` and `DS_Store` need to be removed from the PR.

File: node_modules/oae-authentication/emailTemplates/reset.html.jst [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+        </style>
+    </head>
+    <body>
+        <table cellspacing="10" cellpadding="0">
+            <tbody>
+                <% renderHeader(); %>
+
+                <tr class="row content">
+                    <td>
+                        Please try to reset your password by visiting the following URL:<br />
```
Comment (sallakarppinen) [inline node_modules/oae-authentication/emailTemplates/reset.html.jst]: I'd take out the word 'try' here (insert appropriate Star Wars -quote). :)

File: node_modules/oae-authentication/lib/api.js [hunk 11]
```diff
@@ -16,23 +16,25 @@
...
 var events = require('events');
 var passport = require('passport');
-
+var crypto = require('crypto');
 var AuthzConstants = require('oae-authz/lib/constants').AuthzConstants;
 var Cassandra = require('oae-util/lib/cassandra');
 var ConfigAPI = require('oae-config');
 var Context = require('oae-context').Context;
 var Locking = require('oae-util/lib/locking');
+var EmailAPI = require('oae-email');
```
Comment (sallakarppinen) [inline node_modules/oae-authentication/lib/api.js]: The dependencies should be in alphabetised groups - first external dependencies, then other OAE module dependencies, then dependencies to the same module.

File: etc/migration/addSecretColumnToAuthenticationLoginId.cql [hunk 12]
```diff
@@ -0,0 +1 @@
+ALTER TABLE AuthenticationLoginId ADD secret varchar;
```
Comment (sallakarppinen) [inline etc/migration/addSecretColumnToAuthenticationLoginId.cql]: This new column should also go in `init.js` (and the file should end in a newline).

File: node_modules/oae-authentication/lib/api.js [hunk 13]
```diff
@@ -883,15 +883,14 @@ var resetPassword = module.exports.resetPassword = function(ctx, tenantAlias, us
 
         // If the secret column was not found or its value didn't match, something is wrong.
         if (!dbSecret || dbSecret.value !== secret) {
-            log().warn({'loginId': loginId}, 'Potential hacking attempt detected');
+            log().warn({'loginId': loginId}, 'Someone tried to use an expired password reset secret');
```
Comment (sallakarppinen) [inline node_modules/oae-authentication/lib/api.js]: This is a much better log message. 👍

File: node_modules/oae-authentication/lib/api.js [hunk 14]
```diff
@@ -677,6 +679,113 @@ var getUserIdFromLoginId = module.exports.getUserIdFromLoginId = function(tenant
...
+            } else if (!user.email) {
+                log().warn({'userId': userId}, 'Used asked for password reset but has no email address');
+                return callback({'code': 400, 'msg': 'This user has no email address, you will have to ask an admin to reset the password'});
+            }
+
+            // Generate a secret token
+            var secret = crypto.randomBytes(16).toString('hex');
+
+            // The secret is stored for 24 hours
+            Cassandra.runQuery('UPDATE "AuthenticationLoginId" USING TTL 86400 SET "secret" = ? WHERE "loginId" = ?', [secret, _flattenLoginId(loginId)], function(err) {
```
Comment (stevenxzhou) [inline node_modules/oae-authentication/lib/api.js]: Thanks. I've created a file called addSecretColumnToAuthenticationLoginId.cql to keep the string for adding column `secret` into column family  `AuthenticationLoginId `. Would it be fine?

File: node_modules/oae-authentication/emailTemplates/reset.html.jst [hunk 15]
```diff
@@ -0,0 +1,190 @@
+<%
+    /**
+     * Render the email header
+     */
+    var renderHeader = function() {
+        var institutionalLogoURL = shared.ensureAbsoluteLink(util.ui.getHashedPath(skinVariables['institutional-logo-url'].replace(/\'/g, '')), baseUrl);
```
Comment (stevenxzhou) [inline node_modules/oae-authentication/emailTemplates/reset.html.jst]: Yes. It will load `OAE` logo by default. Let me know if we need to do further checks on this.

File: node_modules/oae-authentication/emailTemplates/reset.html.jst (hunk 2 above)
Comment (stevenxzhou) [inline node_modules/oae-authentication/emailTemplates/reset.html.jst]: Thanks. :)

File: etc/migration/addSecretColumnToAuthenticationLoginId.cql (hunk 12 above)
Comment (stevenxzhou) [inline etc/migration/addSecretColumnToAuthenticationLoginId.cql]: Thanks Done.

**Comment by sallakarppinen** [comment]
All looks good - I'll wait for the front-end to be finished and then merge both PRs together to keep things consistent.

File: etc/migration/addSecretColumnToAuthenticationLoginId.cql [hunk 16]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE AuthenticationLoginId ADD secret text;
```
Comment (sallakarppinen) [inline etc/migration/addSecretColumnToAuthenticationLoginId.cql]: This needs to be `ALTER TABLE "AuthenticationLoginId" ADD "secret" text;`

Comment (stevenxzhou) [inline etc/migration/addSecretColumnToAuthenticationLoginId.cql]: Thanks. Done!

**Comment by sallakarppinen** [comment]
There's still one `.DS_Store` file included, otherwise this is ready to merge IMO. (The path is  `node_modules/oae-authentication/emailTemplates/.DS_Store`)

**Comment by sallakarppinen** [comment]
There's still one `.DS_Store` file in this PR, but I'll merge this and take it out separately.
